File: backend/internal/repositories/householdrepository.go
package repositories

import (
	"keybook/backend/internal/dtos"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/models"
)

type IHouseholdRepository interface {
	GetHouseholdById(id string) (dtos.HouseholdDto, error)
	GetHouseholdsByPropertyId(propertyId string) ([]dtos.HouseholdDto, error)
	CreateHousehold(personId, propertyId string) (dtos.HouseholdDto, error)
	DeleteHousehold(id string) error
}

type HouseholdRepository struct {
	app *pocketbase.PocketBase
}

func NewHouseholdRepository(app *pocketbase.PocketBase) IHouseholdRepository {
	return &HouseholdRepository{app}
}

func (r *HouseholdRepository) GetHouseholdById(id string) (dtos.HouseholdDto, error) {
	record, err := r.app.Dao().FindRecordById("households", id)
	if err != nil {
		return dtos.HouseholdDto{}, err
	}
	return householdRecordToDto(record), nil
}

func (r *HouseholdRepository) GetHouseholdsByPropertyId(propertyId string) ([]dtos.HouseholdDto, error) {
	records, err := r.app.Dao().FindRecordsByExpr("households",
		dbx.NewExp("property = {:propertyId}", dbx.Params{"propertyId": propertyId}))
	if err != nil {
		return nil, err
	}
	result := make([]dtos.HouseholdDto, 0, len(records))
	for _, record := range records {
		result = append(result, householdRecordToDto(record))
	}
	return result, nil
}

func (r *HouseholdRepository) CreateHousehold(personId, propertyId string) (dtos.HouseholdDto, error) {
	col, err := r.app.Dao().FindCollectionByNameOrId("households")
	if err != nil {
		return dtos.HouseholdDto{}, err
	}
	record := models.NewRecord(col)
	record.Set("person", personId)
	record.Set("property", propertyId)
	if err := r.app.Dao().SaveRecord(record); err != nil {
		return dtos.HouseholdDto{}, err
	}
	return householdRecordToDto(record), nil
}

func (r *HouseholdRepository) DeleteHousehold(id string) error {
	record, err := r.app.Dao().FindRecordById("households", id)
	if err != nil {
		return err
	}
	return r.app.Dao().DeleteRecord(record)
}

func householdRecordToDto(r *models.Record) dtos.HouseholdDto {
	return dtos.HouseholdDto{
		Id:       r.GetId(),
		Person:   r.GetString("person"),
		Property: r.GetString("property"),
	}
}

File: frontend/src/lib/repositories/property/household-repository.ts
import PocketBase from 'pocketbase';
import type { IHouseholdModel } from '$lib/models/property-models';

export interface IHouseholdRepository {
  getHouseholdsByPropertyId(propertyId: string): Promise<IHouseholdModel[]>;
  getHouseholdsByPersonId(personId: string): Promise<IHouseholdModel[]>;
  createHousehold(personId: string, propertyId: string): Promise<IHouseholdModel>;
  deleteHousehold(id: string): Promise<void>;
}

export class HouseholdRepository implements IHouseholdRepository {
  constructor(private readonly backendClient: PocketBase) {}

  async getHouseholdsByPropertyId(propertyId: string): Promise<IHouseholdModel[]> {
    const allHouseholds = await this.backendClient.collection('households').getFullList<IHouseholdModel>();
    return allHouseholds.filter((household) => household.property === propertyId);
  }

  async getHouseholdsByPersonId(personId: string): Promise<IHouseholdModel[]> {
    const allHouseholds = await this.backendClient.collection('households').getFullList<IHouseholdModel>();
    return allHouseholds.filter((household) => household.person === personId);
  }

  async createHousehold(personId: string, propertyId: string): Promise<IHouseholdModel> {
    return await this.backendClient
      .collection('households')
      .create<IHouseholdModel>({ person: personId, property: propertyId });
  }

  async deleteHousehold(id: string): Promise<void> {
    await this.backendClient.collection('households').delete(id);
  }
}
